vertical-pod-autoscaler/updater/updater.go

```diff
@@ -0,0 +1,178 @@
+			continue
+		}
+
+		if len(podsList) == 0 {
+			glog.Infof("No pods matching selector %v", selector)
+			continue
+		}
+
+		evictionController := u.evictionFactrory.NewEvictionController(podsList)
+
+		podsForUpdate := u.processPodsRecommendations(podsList, vpaObj, evictionController)
+
+		for _, pod := range podsForUpdate {
+			glog.Infof("Evicting pod %v", pod.Name)
+			evictErr := evictionController.Evict(pod)
+			if evictErr != nil {
+				glog.Warningf("Evicting pod %v failed", pod.Name, evictErr)
+			}
+		}
+		glog.Infof("Processing loop finished")
+	}
+}
+
+// Returns list of pods that should be updated ordered by update priority
+func (u *updater) processPodsRecommendations(pods []*apiv1.Pod, vpaObj *api_mock.VpaObject,
+	evictionController eviction.EvictionController) []*apiv1.Pod {
+
+	priorityCalculator := update_priority.NewUpdatePriorityCalculator(&vpaObj.ResourcesPolicy, nil)
+
+	for _, pod := range pods {
+		glog.Infof("Processing pod : %v", pod.Name)
+
+		if !evictionController.CanEvict(pod) {
+			glog.Infof("Pod not suitable for eviction : %v", pod.Name)
+			continue
+		}
+
+		recommendation, err := u.recommender.Get(&pod.Spec)
+
+		if err != nil {
+			glog.Errorf("Error while getting recommendation for pod %v: %v", pod.Name, err)
+			continue
+		}
+
+		if recommendation == nil {
+			if len(vpaObj.Recommendation.Containers) == 0 {
+				glog.Errorf("No recommendation for pod %v: %v", pod.Name)
+				continue
+			} else {
+				glog.Infof("Fallback to default VPA recommendation for pod : %v", pod.Name)
+				recommendation = &vpaObj.Recommendation
+			}
+		}
+
+		priorityCalculator.AddPod(pod, recommendation)
+	}
+
+	return priorityCalculator.GetSortedPods()
+}
+
+
+func createSelectorForLabels(labelsMap *map[string]string) (labels.Selector, error) {
+	if len(*labelsMap) == 0 {
+		// aways need labels ? or return labels.Everything() ?
+		return nil, fmt.Errorf("No labels for creating selector")
+	}
+	selector := labels.NewSelector()
+	for k, v := range *labelsMap {
+		r, err := labels.NewRequirement(k, selection.Equals, []string{v})
+		if err != nil {
+			return nil, err
+		}
+		selector = selector.Add(*r)
+	}
+	return selector, nil
+}
+
+func NewUpdater(kubeClient kube_client.Interface, cacheTTl time.Duration, minReplicasForEvicition int, evictionToleranceFraction float64) Updater {
+	return &updater{
+		vpaLister:   newVpaLister(kubeClient),
+		podLister:   newPodLister(kubeClient),
+		recommender: recommender.NewCachedRecommender(cacheTTl, api_mock.NewRecommenderAPI()),
+		evictionFactrory: eviction.NewEvictionControllerFactory(kubeClient, minReplicasForEvicition, evictionToleranceFraction),
+	}
+}
+
+func newVpaLister(kubeClient kube_client.Interface) api_mock.VpaLister {
+	return api_mock.NewVpaLister(kubeClient)
+}
+
+func newPodLister(kubeClient kube_client.Interface) v1lister.PodLister {
+	selector := fields.ParseSelectorOrDie("spec.nodeName!=" + "" + ",status.phase!=" +
```

Comment:
Could not parse lister selector with ObjectMeta.DeletionTimestamp. Added filtering at loop start instead